File: Pset1/Credit.c
// Credit

#include <stdio.h>
#include <cs50.h>

char *type_of_valid_credit(int *arr) ;
int *long_to_array(long long int x) ;
int Check_Valid_Or_not(int *arr) ;


int main(void)
{ 
    
    long long int credit_number ;
    int *arr ; 
    int y ; 
    char *result = "INVALID" ;
    
   
    // Get Credit Number From the user 
    credit_number = get_long("Number: ") ;
     
    //credit_number = 4003600000000014;
    
    // Convert The Number To Array
    arr = long_to_array(credit_number) ;

    
    // Check The Number Is Valid Or Not
    if (arr[0] == -1)
    {
        printf("%s\n", result) ;
    }
    else
    {
        y = Check_Valid_Or_not(arr) ;
            
        if (y == 0)
        {
            printf("%s\n", result) ;
        }
        else
        {
            result  = type_of_valid_credit(arr) ;
                
            printf("%s\n", result) ;
        }
         
        
    }
    

    
}

// Determine The Type Of Credit Card
char *type_of_valid_credit(int *arr)
{
    static char *credit ;
    int i;
    
    
    for (i = 15; i >= 0; i--)
    {
        if (arr[i] > 0)
        {
            break;
        }

    }

    // check the type of Credit Card
    if (arr[i] == 4)
    {
        credit = "VISA" ;
    }
    else if (((arr[i] == 3) && (arr[i - 1] == 4)) || ((arr[i] == 3) && (arr[i - 1] == 7)))
    {
        credit = "AMEX" ;
    }
    else if (((arr[i] == 5) && (arr[i - 1] == 1)) || ((arr[i] == 5) && (arr[i - 1] == 2)))
    {
        credit = "MASTERCARD" ;
    }
    else if (((arr[i] == 5) && (arr[i - 1] == 3)) || ((arr[i] == 5) && (arr[i - 1] == 4)))
    {
        credit = "MASTERCARD" ;
    }
    else if ((arr[i] == 5) && (arr[i - 1] == 5))
    {
        credit = "MASTERCARD" ;
    }
    else
    {
        credit = "INVALID" ; 
    }
    
    
    return credit ;
    
}

// Convert the number to array

int *long_to_array(long long int x)
{
    int i = 0 ;
    
    static int arr[16] = {0};
    int *ptr ;
    static int y = -1 ;
    ptr = &y ; 
    
    while (x > 0)
    {
        arr[i] = x % 10 ;
        x = x / 10 ;
        i++ ;
    }
    
    if (i < 13)
    {
        return ptr ;
    }
    else
    {
        return arr ;  
    }
   
    
}

// Check The Credit Is Valid Or Not
int Check_Valid_Or_not(int *arr)
{

    int count = 0 ;
    int i ;
    int temp ;

    for (i = 1; i < 16; i = i + 2)
    {

        temp = arr[i] * 2;

        while (temp > 0)
        {
            count += temp % 10;
            temp = temp / 10 ;
        }

    }
   
    for (i = 0; i < 16; i = i + 2)
    {

        count += arr[i]  ;

    }
 
    if ((count % 10) == 0)
    {
        return 1 ;
    }
    else
    {
        return 0 ;
    }

}

File: pset6 2020/credit.py
def main():

    number = []
    number = get_input()

    # check the number is valid or not
    state = valid_or_not(number)

    if state == 0:
        print("INVALID")
        return 1

    # if the number is valid determine the type of credit
    credit = type_of_credit(number)

    print(credit)

# function to get input from the user


def get_input():

    number = []

    # loop until you get a valid input
    while (True):
        count = 0
        # get a number from user
        number = input("Number: ")
        # check if the number contain only digits
        for i in range(len(number)):
            # check if there is anythine other than a digit
            if number[i].isdigit() == False:
                count += 1
                break
        # if the number is valid
        if count == 0:
            break

    # return the number
    return number


# function to check the input is valid or not
def valid_or_not(number):

    sum = 0
    # Multiply every other digit by 2, starting with the number’s second-to-last digit
    # and then add those products’ digits together.
    for i in range(len(number) - 2, -1, -2):

        temp = 2 * int(number[i])

        while (temp > 0):

            sum += int(temp % 10)
            temp = int(temp / 10)

    # dd the sum to the sum of the digits that weren’t multiplied by 2.
    for i in range(len(number) - 1, -1, -2):

        sum += int(number[i])

    # check the number is valid or not
    # the number is valid if the last digit is 0
    if sum % 10 == 0:
        return 1
    else:
        return 0


# function to determine the type of credit
def type_of_credit(number):

    # check the type of Credit Card
    if number[0] == '4':
        credit = "VISA"
    elif number[0] == '3' and number[1] == '4':
        credit = "AMEX"
    elif number[0] == '3' and number[1] == '7':
        credit = "AMEX"
    elif number[0] == '5' and number[1] == '1':
        credit = "MASTERCARD"
    elif number[0] == '5' and number[1] == '2':
        credit = "MASTERCARD"
    elif number[0] == '5' and number[1] == '3':
        credit = "MASTERCARD"
    elif number[0] == '5' and number[1] == '4':
        credit = "MASTERCARD"
    elif number[0] == '5' and number[1] == '5':
        credit = "MASTERCARD"
    else:
        credit = "INVALID"

    return credit


main()